tasks/utils/oauth.ts

```diff
@@ -0,0 +1,151 @@
+import { getConfig } from './config'
+import type { AuthorizationServer, Client } from 'oauth4webapi'
+import * as oauth from 'oauth4webapi'
+import {
+  authorizationCodeGrantRequest,
+  discoveryRequest,
+  generateRandomState,
+  isOAuth2Error,
+  processAuthorizationCodeOpenIDResponse,
+  processDiscoveryResponse,
+  validateAuthResponse
+} from 'oauth4webapi'
+import { getCookie, setCookie } from 'cookies-next'
+
+const config = getConfig()
+const codeChallengeMethod = 'S256'
+
+const LOCALSTORAGE_KEY_STATE = 'oauth_state'
+const LOCALSTORAGE_KEY_CODE_VERIFIER = 'oauth_code_verifier'
+
+const getAuthorizationServer = async (issuerUrl: string): Promise<AuthorizationServer> => {
+  const issuer = new URL(issuerUrl)
+  const authorizationServer = await discoveryRequest(issuer)
+    .then((res) => processDiscoveryResponse(issuer, res))
+
+  if (!authorizationServer.code_challenge_methods_supported?.includes(codeChallengeMethod)) {
+    throw new Error('Authorization server does not support required code challenge')
+  }
+
+  return authorizationServer
+}
+
+const getCommonOAuthEntities = async (overrideConfig?: Partial<typeof config.oauth>): Promise<{ authorizationServer: AuthorizationServer, client: Client }> => {
+  const { issuerUrl, clientId } = { ...config.oauth, ...overrideConfig }
+
+  const authorizationServer = await getAuthorizationServer(issuerUrl)
+  const client: Client = {
+    client_id: clientId,
+    token_endpoint_auth_method: 'none',
+  }
+
+  return { authorizationServer, client }
+}
+
+const storeValue = (key: string, value: string) => {
+  const expiresAt = new Date()
+  expiresAt.setHours(expiresAt.getHours() + 1)
+  const secureCookie = config.env === 'production'
+  setCookie(key, value, { expires: expiresAt, sameSite: 'lax', secure: secureCookie })
```

Comment:
Can we move this management to the backend?
We should try to avoid handling cookies on the frontend.
There is this cool option called ` samesite=strict` which disables javascript touching this cookie.